File: python-checks/src/main/java/org/sonar/python/checks/DeprecatedNumpyTypesCheck.java
/*
 * SonarQube Python Plugin
 * Copyright (C) 2011-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
package org.sonar.python.checks;

import java.util.Map;
import java.util.Optional;
import org.sonar.check.Rule;
import org.sonar.plugins.python.api.PythonSubscriptionCheck;
import org.sonar.plugins.python.api.SubscriptionContext;
import org.sonar.plugins.python.api.quickfix.PythonQuickFix;
import org.sonar.plugins.python.api.symbols.Symbol;
import org.sonar.plugins.python.api.tree.QualifiedExpression;
import org.sonar.plugins.python.api.tree.Tree;
import org.sonar.python.quickfix.TextEditUtils;

@Rule(key = "S6730")
public class DeprecatedNumpyTypesCheck extends PythonSubscriptionCheck {

  private static final String MESSAGE = "Replace this deprecated \"numpy\" type alias with the builtin type \"%s\".";
  private static final String QUICK_FIX_MESSAGE = "Replace with %s.";
  private static final Map<String, String> TYPE_TO_CHECK = Map.of(
      "numpy.bool", "bool",
      "numpy.int", "int",
      "numpy.float", "float",
      "numpy.complex", "complex",
      "numpy.object", "object",
      "numpy.str", "str",
      "numpy.long", "int",
      "numpy.unicode", "str");

  @Override
  public void initialize(Context context) {
    context.registerSyntaxNodeConsumer(Tree.Kind.QUALIFIED_EXPR,
        DeprecatedNumpyTypesCheck::checkForDeprecatedTypesNames);
  }

  private static void checkForDeprecatedTypesNames(SubscriptionContext ctx) {
    QualifiedExpression expression = (QualifiedExpression) ctx.syntaxNode();
    Optional.ofNullable(expression.symbol())
        .map(Symbol::fullyQualifiedName)
        .map(TYPE_TO_CHECK::get)
        .ifPresent(type -> raiseIssue(expression, type, ctx));
  }

  private static void raiseIssue(Tree expression, String replacementType, SubscriptionContext ctx) {
    PreciseIssue issue = ctx.addIssue(expression, String.format(MESSAGE, replacementType));
    PythonQuickFix quickFix = PythonQuickFix.newQuickFix(
        String.format(QUICK_FIX_MESSAGE, replacementType),
        TextEditUtils.replace(expression, replacementType));
    issue.addQuickFix(quickFix);
  }
}

File: python-checks/src/test/resources/checks/deprecatedNumpyTypesCheck.py
import numpy as np


def failure():
    b = np.bool(True)  # Noncompliant {{Replace this deprecated "numpy" type alias with the builtin type "bool".}}
    #   ^^^^^^^
    i = np.int(42)  # Noncompliant {{Replace this deprecated "numpy" type alias with the builtin type "int".}}
    #   ^^^^^^
    f = np.float(4.2)  # Noncompliant {{Replace this deprecated "numpy" type alias with the builtin type "float".}}
    #   ^^^^^^^^
    c = np.complex(-2.0, 0.0)  # Noncompliant {{Replace this deprecated "numpy" type alias with the builtin type "complex".}}
    #   ^^^^^^^^^^
    o = np.object  # Noncompliant {{Replace this deprecated "numpy" type alias with the builtin type "object".}}
    #   ^^^^^^^^^
    s = np.str("foo")  # Noncompliant {{Replace this deprecated "numpy" type alias with the builtin type "str".}}
    #   ^^^^^^
    l = np.long(123)  # Noncompliant {{Replace this deprecated "numpy" type alias with the builtin type "int".}}
    #   ^^^^^^^
    u = np.unicode("bar")  # Noncompliant {{Replace this deprecated "numpy" type alias with the builtin type "str".}}
    #   ^^^^^^^^^^

def failure_import_as_z():
    import numpy as z
    b = z.bool(True)  # Noncompliant
    #   ^^^^^^
    i = z.int(42)  # Noncompliant
    #   ^^^^^
    f = z.float(4.2)  # Noncompliant
    #   ^^^^^^^
    c = z.complex(-2.0, 0.0)  # Noncompliant
    #   ^^^^^^^^^
    o = z.object  # Noncompliant
    #   ^^^^^^^^
    s = z.str("foo")  # Noncompliant
    #   ^^^^^
    l = z.long(123)  # Noncompliant
    #   ^^^^^^
    u = z.unicode("bar")  # Noncompliant
    #   ^^^^^^^^^

def success():
    b = True  # Compliant
    b = bool(True)  # Compliant
    i = 42  # Compliant
    i = int(42)  # Compliant
    f = 4.2  # Compliant
    f = float(4.2)  # Compliant
    c = complex(-2.0, 0.0)  # Compliant
    o = object  # Compliant
    s = "foo"  # Compliant
    s = str("foo")  # Compliant
    l = 123  # Compliant
    l = int(123)  # Compliant
    u = "bar"  # Compliant
    u = str("bar")  # Compliant

